{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ollama\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import time\n",
    "import pickle\n",
    "import random\n",
    "import os\n",
    "import requests\n",
    "import json\n",
    "import logging\n",
    "\n",
    "\n",
    "def get_response(prompt):\n",
    "    url = \"https://chatapi.midjourney-vip.cn/v1/chat/completions\"\n",
    "    model = \"gpt-4o-2024-05-13\"\n",
    "    payload = json.dumps({\n",
    "    \"model\": model,\n",
    "    \"messages\": [\n",
    "        {\n",
    "            \"role\": \"user\",\n",
    "            \"content\": prompt\n",
    "        }\n",
    "    ]})\n",
    "    headers = {\n",
    "    'Accept': 'application/json',\n",
    "    'Authorization': 'sk-OOLfa5EgvvQNeXP03953498fE4004dB3811e10B582C2C127',\n",
    "    'User-Agent': 'Apifox/1.0.0 (https://apifox.com)',\n",
    "    'Content-Type': 'application/json'\n",
    "    }\n",
    "    response = requests.request(\"POST\", url, headers=headers, data=payload)\n",
    "\n",
    "    responses = response.text\n",
    "    return responses\n",
    "\n",
    "def is_valid_output(feature):\n",
    "    return len(feature.split()) < 8 and feature != ''\n",
    "\n",
    "def get_prompt(finetune_prompt):\n",
    "    number = 0\n",
    "    while True:\n",
    "        response = get_response(finetune_prompt)\n",
    "        # print(response)\n",
    "        \n",
    "        prompt_positions = [i for i in range(len(response)) if response.startswith('<PROMPT>', i)]\n",
    "        \n",
    "        if len(prompt_positions) < 2:\n",
    "            print(\"Invalid output format, retrying...\")\n",
    "            number = number+1\n",
    "            if number > 8:\n",
    "                return None\n",
    "            continue\n",
    "\n",
    "        begin = prompt_positions[-2] + len('<PROMPT>\"')\n",
    "        end = prompt_positions[-1]\n",
    "\n",
    "        response = response[begin:end].replace('\\\\n', '\\n')\n",
    "\n",
    "        begin = response.find('Task')\n",
    "        end = response.rfind('Gene_name') + len('Gene_name')\n",
    "        response = response[begin:end]\n",
    "\n",
    "        return response\n",
    "\n",
    "        \n",
    "def compare_list(word_list):\n",
    "    common_features = word_list[0]\n",
    "    number = 0\n",
    "    for lst in word_list[1:]:\n",
    "        number += len(set(common_features) & set(lst))\n",
    "\n",
    "    return number\n",
    "\n",
    "def get_num():\n",
    "    if not os.path.exists('txt/data/good_num.npy'):\n",
    "        if not os.path.exists('txt/data/ori.npy'):\n",
    "            with open('/home/yuantao/code/LLM/txt/prompt/prompt.txt', 'r') as f:\n",
    "                prompt = f.read()\n",
    "                ori_num = []\n",
    "                for idx, gene_name in enumerate(gene_name_list):\n",
    "                    word_list = [] \n",
    "                    for i in range(4):\n",
    "                        \n",
    "                        while True:\n",
    "                            new_prompt = prompt.replace('Gene_name', gene_name)\n",
    "                            content = ollama.generate(model='gemma2', prompt=new_prompt)['response']\n",
    "\n",
    "                            start_pos = content.find('<begin>')\n",
    "                            end_pos = content.rfind('<end>')\n",
    "                            \n",
    "                            if start_pos == -1 or end_pos == -1 or start_pos >= end_pos:\n",
    "                                print(\"Invalid output format, retrying...\")\n",
    "                                continue\n",
    "\n",
    "                            feature_start_idx = start_pos + len('<begin>')\n",
    "                            feature_end_idx = end_pos\n",
    "                            feature = content[feature_start_idx:feature_end_idx].strip()\n",
    "                            \n",
    "\n",
    "                            feature_list = feature.split(',')\n",
    "                            if len(feature_list) >= MAX_FEATURES: \n",
    "                                feature_list = feature_list[:MAX_FEATURES]\n",
    "                            \n",
    "                            if is_valid_output(feature_list[0]):\n",
    "                                feature_list = [word.strip().lower() for word in feature_list if word != '']\n",
    "                                word_list.append(feature_list)\n",
    "                                break\n",
    "                            else:\n",
    "                                print(\"Invalid feature\")\n",
    "                                print(\"Retrying...\")\n",
    "\n",
    "                    count = compare_list(word_list)\n",
    "                    print(gene_name, count)\n",
    "                    ori_num.append(count)\n",
    "\n",
    "                print('ori_num:', ori_num)\n",
    "                np.save('txt/data/ori.npy', ori_num)\n",
    "                good_num = ori_num\n",
    "                bad_num = ori_num\n",
    "        else:\n",
    "            good_num = np.load('txt/data/ori.npy')\n",
    "            bad_num = np.load('txt/data/ori.npy')\n",
    "    else:\n",
    "        good_num = np.load('txt/data/good_num.npy')\n",
    "        bad_num = np.load('txt/data/bad_num.npy')\n",
    "\n",
    "    # print(good_num,np.mean(good_num))\n",
    "    # print(bad_num,np.mean(bad_num))\n",
    "    return good_num, bad_num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSG00000179627,ZBTB42', 'ENSG00000154269,ENPP3', 'ENSG00000164096,C4orf3', 'ENSG00000188191,PRKAR1B', 'ENSG00000131779,PEX11B', 'ENSG00000151164,RAD9B', 'ENSG00000162670,BRINP3', 'ENSG00000171848,RRM2', 'ENSG00000174132,FAM174A', 'ENSG00000103546,SLC6A2', 'ENSG00000115762,PLEKHB2', 'ENSG00000142082,SIRT3', 'ENSG00000128335,APOL2', 'ENSG00000134897,BIVM', 'ENSG00000167941,SOST', 'ENSG00000119630,PGF', 'ENSG00000102575,ACP5', 'ENSG00000135919,SERPINE2', 'ENSG00000092847,AGO1', 'ENSG00000100302,RASD2']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(array([ 5,  4,  1,  6,  5,  3,  3, 10,  0,  1,  4,  3,  0,  0,  5,  5,  2,\n",
       "         8,  5,  3]),\n",
       " array([ 5,  4,  1,  6,  5,  3,  3, 10,  0,  1,  4,  3,  0,  0,  5,  5,  2,\n",
       "         8,  5,  3]))"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prompt = open(\"txt/prompt/prompt.txt\", \"r\").read()\n",
    "fine_tune_prompt = open(\"txt/prompt/fine_tune_prompt.txt\", \"r\").read()\n",
    "cancer_name = 'pan-cancer'\n",
    "prompt = prompt.replace('Cancer_name', cancer_name)\n",
    "\n",
    "# gene_name_list = open(r'/home/yuantao/code/LLM/txt/data/node_names.txt','r',encoding='utf-8').read().split('\\n')\n",
    "# gene_name_list = random.sample(gene_name_list, 20)\n",
    "gene_name_list = ['ENSG00000179627,ZBTB42', 'ENSG00000154269,ENPP3', 'ENSG00000164096,C4orf3', 'ENSG00000188191,PRKAR1B', 'ENSG00000131779,PEX11B', 'ENSG00000151164,RAD9B', 'ENSG00000162670,BRINP3', 'ENSG00000171848,RRM2', 'ENSG00000174132,FAM174A', 'ENSG00000103546,SLC6A2', 'ENSG00000115762,PLEKHB2', 'ENSG00000142082,SIRT3', 'ENSG00000128335,APOL2', 'ENSG00000134897,BIVM', 'ENSG00000167941,SOST', 'ENSG00000119630,PGF', 'ENSG00000102575,ACP5', 'ENSG00000135919,SERPINE2', 'ENSG00000092847,AGO1', 'ENSG00000100302,RASD2']\n",
    "print(gene_name_list)\n",
    "\n",
    "model = 'gemma2'\n",
    "MAX_FEATURES = 6\n",
    "get_num()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis:\n",
      "Upon comparison, the stable prompt exhibits a slight edge over the suboptimal prompt in terms of repetition accuracy and output stability (number of repeated words on 20 genes being 4.4 vs. 4.2). The stability in the prompt primarily arises from the additional specificity and iterative reinforcement of vocabulary consistency.\n",
      "   \n",
      "Key areas of stability in the stable prompt include:\n",
      "1. Explicit instructions on gene description first and vocabulary summarization next.\n",
      "2. Clear emphasis on aligning descriptive vocabulary directly with described functions and roles.\n",
      "3. Specific mention of avoiding overly common terms, ensuring relevance to cancer biology.\n",
      "4. Instruction to iteratively review vocabulary for alignment, enhancing consistency.\n",
      "    \n",
      "To further boost the number of repetitions and stability, consider:\n",
      "1. Emphasizing the use of specific cancer-related concepts and technical terms.\n",
      "2. Reinforcing unique biological features related to gene functions to reduce variability. \n",
      "3. Systematically reinforcing model consistency through distinct repetitive cues in vocabulary generation.\n",
      "\n",
      "Fine-Tuned Prompt:\n",
      "<Task: Write six descriptive words about the biology of a given gene in a specific cancer.\n",
      "\n",
      "Below is an example of output:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: ENSG00000121879,PIK3CA\n",
      "<description>\n",
      "PIK3CA is a gene that encodes the catalytic subunit p110α of the enzyme phosphatidylinositol 3-kinase (PI3K), which is crucial in the PI3K/AKT signaling pathway, regulating cell growth, proliferation, survival, and metabolism. Located on chromosome 3q26.32, PIK3CA mutations are frequently observed in various cancers, such as breast, colorectal, and endometrial cancers, where they cause constitutive activation of the PI3K/AKT pathway, driving tumor progression. The most common mutations occur in exons 9 and 20, affecting the helical and kinase domains. Besides cancer, these mutations are also involved in overgrowth syndromes like PIK3CA-Related Overgrowth Spectrum (PROS). Due to its critical role in oncogenesis, PIK3CA is a significant therapeutic target, with PI3K inhibitors like alpelisib used in treating cancers harboring PIK3CA mutations. The gene is also included in genomic profiling to guide personalized cancer therapies.\n",
      "<description>\n",
      "Output:\n",
      "<begin>\n",
      "PI3K/AKT signaling,phosphoinositide phosphorylation,oncogenesis,cell proliferation,metabolic regulation,protein kinase activation\n",
      "<end>\n",
      "\n",
      "Requirements:\n",
      "1. The description of the gene should focus on the known gene information features, including but not limited to known literature and reports. For some unknown information, you should try to complete them with your knowledge, and select biological features related to the gene, such as function, location, interaction, regulatory mechanism, etc. Consider the mechanism of action of the gene in cancer.\n",
      "2. You need to think about how to describe this gene step by step. Output the description paragraph first, and then output the summary.\n",
      "3. Your description needs to include the biological function of the gene and information related to the given disease, but the disease name is not allowed in the output. Note: not all genes are related to this disease. Please carefully confirm the source and give me the relevant description. If this gene has no information related to this disease, please only introduce the role of this gene to me.\n",
      "4. Output the description <description>...<description> first, then summarize the characteristic vocabulary from the description.\n",
      "5. The characteristic vocabulary format is <begin>,,,<end>. For example: <begin>PI3K/AKT signaling, phosphoinositide phosphorylation, oncogenesis, cell proliferation, metabolic regulation, protein kinase activation<end>\n",
      "6. Ensure the vocabulary is unique and directly aligned with the described functions and roles of the gene in cancer biology. Highlight the significant differences from other genes.\n",
      "7. Take care to minimize repetitions in the descriptive vocabulary. Steer clear of outputting overly common terms that are less specific to the gene’s mechanics in cancer biology.\n",
      "8. You should incorporate any repetitive aspects intentionally to ensure consistency in the output. Reiteratively review the vocabulary for alignment with gene-specific biological features.\n",
      "9. Reinforce distinctive cancer-related concepts and unique biological terms to ensure precise alignment in gene-specific vocabulary generation.\n",
      "\n",
      "Task:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: Gene_name\n",
      "ENSG00000179627,ZBTB42 5\n",
      "ENSG00000154269,ENPP3 4\n",
      "ENSG00000164096,C4orf3 4\n",
      "ENSG00000188191,PRKAR1B 9\n",
      "ENSG00000131779,PEX11B 6\n",
      "ENSG00000151164,RAD9B 2\n",
      "Invalid output format, retrying...\n",
      "ENSG00000162670,BRINP3 3\n",
      "ENSG00000171848,RRM2 7\n",
      "ENSG00000174132,FAM174A 1\n",
      "ENSG00000103546,SLC6A2 3\n",
      "ENSG00000115762,PLEKHB2 7\n",
      "ENSG00000142082,SIRT3 3\n",
      "ENSG00000128335,APOL2 3\n",
      "ENSG00000134897,BIVM 0\n",
      "ENSG00000167941,SOST 2\n",
      "ENSG00000119630,PGF 3\n",
      "ENSG00000102575,ACP5 5\n",
      "ENSG00000135919,SERPINE2 4\n",
      "ENSG00000092847,AGO1 7\n",
      "ENSG00000100302,RASD2 6\n",
      "GPT_num: [5, 4, 4, 9, 6, 2, 3, 7, 1, 3, 7, 3, 3, 0, 2, 3, 5, 4, 7, 6]\n",
      "GPT_num: 4.2\n",
      "GPT is worse than good and bad\n",
      "Analysis:\n",
      "Upon comparison, the stable prompt exhibits a slight edge over the suboptimal prompt in terms of repetition accuracy and output stability (number of repeated words on 20 genes being 4.4 vs. 4.2). The stability in the prompt primarily arises from the additional specificity and iterative reinforcement of vocabulary consistency.\n",
      "   \n",
      "Key areas of stability in the stable prompt include:\n",
      "1. Explicit instructions on gene description first and vocabulary summarization next.\n",
      "2. Clear emphasis on aligning descriptive vocabulary directly with described functions and roles.\n",
      "3. Specific mention of avoiding overly common terms, ensuring relevance to cancer biology.\n",
      "4. Instruction to iteratively review vocabulary for alignment, enhancing consistency.\n",
      "    \n",
      "To further boost the number of repetitions and stability, consider:\n",
      "1. Emphasizing the use of specific cancer-related concepts and technical terms.\n",
      "2. Reinforcing unique biological features related to gene functions to reduce variability. \n",
      "3. Systematically reinforcing model consistency through distinct repetitive cues in vocabulary generation.\n",
      "\n",
      "Fine-Tuned Prompt:\n",
      "<Task: Write six descriptive words about the biology of a given gene in a specific cancer.\n",
      "\n",
      "Below is an example of output:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: ENSG00000121879,PIK3CA\n",
      "<description>\n",
      "PIK3CA is a gene that encodes the catalytic subunit p110α of the enzyme phosphatidylinositol 3-kinase (PI3K), which is crucial in the PI3K/AKT signaling pathway, regulating cell growth, proliferation, survival, and metabolism. Located on chromosome 3q26.32, PIK3CA mutations are frequently observed in various cancers, such as breast, colorectal, and endometrial cancers, where they cause constitutive activation of the PI3K/AKT pathway, driving tumor progression. The most common mutations occur in exons 9 and 20, affecting the helical and kinase domains. Besides cancer, these mutations are also involved in overgrowth syndromes like PIK3CA-Related Overgrowth Spectrum (PROS). Due to its critical role in oncogenesis, PIK3CA is a significant therapeutic target, with PI3K inhibitors like alpelisib used in treating cancers harboring PIK3CA mutations. The gene is also included in genomic profiling to guide personalized cancer therapies.\n",
      "<description>\n",
      "Output:\n",
      "<begin>\n",
      "PI3K/AKT signaling,phosphoinositide phosphorylation,oncogenesis,cell proliferation,metabolic regulation,protein kinase activation\n",
      "<end>\n",
      "\n",
      "Requirements:\n",
      "1. The description of the gene should focus on the known gene information features, including but not limited to known literature and reports. For some unknown information, you should try to complete them with your knowledge, and select biological features related to the gene, such as function, location, interaction, regulatory mechanism, etc. Consider the mechanism of action of the gene in cancer.\n",
      "2. You need to think about how to describe this gene step by step. Output the description paragraph first, and then output the summary.\n",
      "3. Your description needs to include the biological function of the gene and information related to the given disease, but the disease name is not allowed in the output. Note: not all genes are related to this disease. Please carefully confirm the source and give me the relevant description. If this gene has no information related to this disease, please only introduce the role of this gene to me.\n",
      "4. Output the description <description>...<description> first, then summarize the characteristic vocabulary from the description.\n",
      "5. The characteristic vocabulary format is <begin>,,,<end>. For example: <begin>PI3K/AKT signaling, phosphoinositide phosphorylation, oncogenesis, cell proliferation, metabolic regulation, protein kinase activation<end>\n",
      "6. Ensure the vocabulary is unique and directly aligned with the described functions and roles of the gene in cancer biology. Highlight the significant differences from other genes.\n",
      "7. Take care to minimize repetitions in the descriptive vocabulary. Steer clear of outputting overly common terms that are less specific to the gene’s mechanics in cancer biology.\n",
      "8. You should incorporate any repetitive aspects intentionally to ensure consistency in the output. Reiteratively review the vocabulary for alignment with gene-specific biological features.\n",
      "9. Reinforce distinctive cancer-related concepts and unique biological terms to ensure precise alignment in gene-specific vocabulary generation.\n",
      "\n",
      "Task:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: Gene_name\n",
      "ENSG00000179627,ZBTB42 8\n",
      "ENSG00000154269,ENPP3 6\n",
      "ENSG00000164096,C4orf3 2\n",
      "ENSG00000188191,PRKAR1B 6\n",
      "ENSG00000131779,PEX11B 6\n",
      "ENSG00000151164,RAD9B 3\n",
      "ENSG00000162670,BRINP3 6\n",
      "ENSG00000171848,RRM2 8\n",
      "ENSG00000174132,FAM174A 2\n",
      "ENSG00000103546,SLC6A2 2\n",
      "ENSG00000115762,PLEKHB2 7\n",
      "ENSG00000142082,SIRT3 8\n",
      "ENSG00000128335,APOL2 4\n",
      "ENSG00000134897,BIVM 0\n",
      "ENSG00000167941,SOST 8\n",
      "ENSG00000119630,PGF 6\n",
      "ENSG00000102575,ACP5 8\n",
      "ENSG00000135919,SERPINE2 2\n",
      "Invalid output format, retrying...\n",
      "ENSG00000092847,AGO1 3\n",
      "ENSG00000100302,RASD2 1\n",
      "GPT_num: [8, 6, 2, 6, 6, 3, 6, 8, 2, 2, 7, 8, 4, 0, 8, 6, 8, 2, 3, 1]\n",
      "GPT_num: 4.8\n",
      "GPT is better than good\n",
      "Analysis:\n",
      "Upon comparison, the stable prompt exhibits a slight edge over the suboptimal prompt in terms of repetition accuracy and output stability (number of repeated words on 20 genes being 4.4 vs. 4.2). The stability in the prompt primarily arises from the additional specificity and iterative reinforcement of vocabulary consistency.\n",
      "   \n",
      "Key areas of stability in the stable prompt include:\n",
      "1. Explicit instructions on gene description first and vocabulary summarization next.\n",
      "2. Clear emphasis on aligning descriptive vocabulary directly with described functions and roles.\n",
      "3. Specific mention of avoiding overly common terms, ensuring relevance to cancer biology.\n",
      "4. Instruction to iteratively review vocabulary for alignment, enhancing consistency.\n",
      "    \n",
      "To further boost the number of repetitions and stability, consider:\n",
      "1. Emphasizing the use of specific cancer-related concepts and technical terms.\n",
      "2. Reinforcing unique biological features related to gene functions to reduce variability. \n",
      "3. Systematically reinforcing model consistency through distinct repetitive cues in vocabulary generation.\n",
      "\n",
      "Fine-Tuned Prompt:\n",
      "<Task: Write six descriptive words about the biology of a given gene in a specific cancer.\n",
      "\n",
      "Below is an example of output:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: ENSG00000121879,PIK3CA\n",
      "<description>\n",
      "PIK3CA is a gene that encodes the catalytic subunit p110α of the enzyme phosphatidylinositol 3-kinase (PI3K), which is crucial in the PI3K/AKT signaling pathway, regulating cell growth, proliferation, survival, and metabolism. Located on chromosome 3q26.32, PIK3CA mutations are frequently observed in various cancers, such as breast, colorectal, and endometrial cancers, where they cause constitutive activation of the PI3K/AKT pathway, driving tumor progression. The most common mutations occur in exons 9 and 20, affecting the helical and kinase domains. Besides cancer, these mutations are also involved in overgrowth syndromes like PIK3CA-Related Overgrowth Spectrum (PROS). Due to its critical role in oncogenesis, PIK3CA is a significant therapeutic target, with PI3K inhibitors like alpelisib used in treating cancers harboring PIK3CA mutations. The gene is also included in genomic profiling to guide personalized cancer therapies.\n",
      "<description>\n",
      "Output:\n",
      "<begin>\n",
      "PI3K/AKT signaling,phosphoinositide phosphorylation,oncogenesis,cell proliferation,metabolic regulation,protein kinase activation\n",
      "<end>\n",
      "\n",
      "Requirements:\n",
      "1. The description of the gene should focus on the known gene information features, including but not limited to known literature and reports. For some unknown information, you should try to complete them with your knowledge, and select biological features related to the gene, such as function, location, interaction, regulatory mechanism, etc. Consider the mechanism of action of the gene in cancer.\n",
      "2. You need to think about how to describe this gene step by step. Output the description paragraph first, and then output the summary.\n",
      "3. Your description needs to include the biological function of the gene and information related to the given disease, but the disease name is not allowed in the output. Note: not all genes are related to this disease. Please carefully confirm the source and give me the relevant description. If this gene has no information related to this disease, please only introduce the role of this gene to me.\n",
      "4. Output the description <description>...<description> first, then summarize the characteristic vocabulary from the description.\n",
      "5. The characteristic vocabulary format is <begin>,,,<end>. For example: <begin>PI3K/AKT signaling, phosphoinositide phosphorylation, oncogenesis, cell proliferation, metabolic regulation, protein kinase activation<end>\n",
      "6. Ensure the vocabulary is unique and directly aligned with the described functions and roles of the gene in cancer biology. Highlight the significant differences from other genes.\n",
      "7. Take care to minimize repetitions in the descriptive vocabulary. Steer clear of outputting overly common terms that are less specific to the gene’s mechanics in cancer biology.\n",
      "8. You should incorporate any repetitive aspects intentionally to ensure consistency in the output. Reiteratively review the vocabulary for alignment with gene-specific biological features.\n",
      "9. Reinforce distinctive cancer-related concepts and unique biological terms to ensure precise alignment in gene-specific vocabulary generation.\n",
      "\n",
      "Task:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: Gene_name\n",
      "ENSG00000179627,ZBTB42 6\n",
      "ENSG00000154269,ENPP3 2\n",
      "ENSG00000164096,C4orf3 1\n",
      "ENSG00000188191,PRKAR1B 7\n",
      "ENSG00000131779,PEX11B 7\n",
      "ENSG00000151164,RAD9B 7\n",
      "ENSG00000162670,BRINP3 2\n",
      "ENSG00000171848,RRM2 2\n",
      "ENSG00000174132,FAM174A 4\n",
      "ENSG00000103546,SLC6A2 3\n",
      "ENSG00000115762,PLEKHB2 5\n",
      "ENSG00000142082,SIRT3 7\n",
      "ENSG00000128335,APOL2 2\n",
      "ENSG00000134897,BIVM 1\n",
      "ENSG00000167941,SOST 3\n",
      "ENSG00000119630,PGF 5\n",
      "ENSG00000102575,ACP5 4\n",
      "ENSG00000135919,SERPINE2 4\n",
      "ENSG00000092847,AGO1 3\n",
      "ENSG00000100302,RASD2 2\n",
      "GPT_num: [6, 2, 1, 7, 7, 7, 2, 2, 4, 3, 5, 7, 2, 1, 3, 5, 4, 4, 3, 2]\n",
      "GPT_num: 3.85\n",
      "GPT is worse than good and bad\n",
      "Analysis:\n",
      "Upon comparison, the stable prompt exhibits a slight edge over the suboptimal prompt in terms of repetition accuracy and output stability (number of repeated words on 20 genes being 4.4 vs. 4.2). The stability in the prompt primarily arises from the additional specificity and iterative reinforcement of vocabulary consistency.\n",
      "   \n",
      "Key areas of stability in the stable prompt include:\n",
      "1. Explicit instructions on gene description first and vocabulary summarization next.\n",
      "2. Clear emphasis on aligning descriptive vocabulary directly with described functions and roles.\n",
      "3. Specific mention of avoiding overly common terms, ensuring relevance to cancer biology.\n",
      "4. Instruction to iteratively review vocabulary for alignment, enhancing consistency.\n",
      "    \n",
      "To further boost the number of repetitions and stability, consider:\n",
      "1. Emphasizing the use of specific cancer-related concepts and technical terms.\n",
      "2. Reinforcing unique biological features related to gene functions to reduce variability. \n",
      "3. Systematically reinforcing model consistency through distinct repetitive cues in vocabulary generation.\n",
      "\n",
      "Fine-Tuned Prompt:\n",
      "<Task: Write six descriptive words about the biology of a given gene in a specific cancer.\n",
      "\n",
      "Below is an example of output:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: ENSG00000121879,PIK3CA\n",
      "<description>\n",
      "PIK3CA is a gene that encodes the catalytic subunit p110α of the enzyme phosphatidylinositol 3-kinase (PI3K), which is crucial in the PI3K/AKT signaling pathway, regulating cell growth, proliferation, survival, and metabolism. Located on chromosome 3q26.32, PIK3CA mutations are frequently observed in various cancers, such as breast, colorectal, and endometrial cancers, where they cause constitutive activation of the PI3K/AKT pathway, driving tumor progression. The most common mutations occur in exons 9 and 20, affecting the helical and kinase domains. Besides cancer, these mutations are also involved in overgrowth syndromes like PIK3CA-Related Overgrowth Spectrum (PROS). Due to its critical role in oncogenesis, PIK3CA is a significant therapeutic target, with PI3K inhibitors like alpelisib used in treating cancers harboring PIK3CA mutations. The gene is also included in genomic profiling to guide personalized cancer therapies.\n",
      "<description>\n",
      "Output:\n",
      "<begin>\n",
      "PI3K/AKT signaling,phosphoinositide phosphorylation,oncogenesis,cell proliferation,metabolic regulation,protein kinase activation\n",
      "<end>\n",
      "\n",
      "Requirements:\n",
      "1. The description of the gene should focus on the known gene information features, including but not limited to known literature and reports. For some unknown information, you should try to complete them with your knowledge, and select biological features related to the gene, such as function, location, interaction, regulatory mechanism, etc. Consider the mechanism of action of the gene in cancer.\n",
      "2. You need to think about how to describe this gene step by step. Output the description paragraph first, and then output the summary.\n",
      "3. Your description needs to include the biological function of the gene and information related to the given disease, but the disease name is not allowed in the output. Note: not all genes are related to this disease. Please carefully confirm the source and give me the relevant description. If this gene has no information related to this disease, please only introduce the role of this gene to me.\n",
      "4. Output the description <description>...<description> first, then summarize the characteristic vocabulary from the description.\n",
      "5. The characteristic vocabulary format is <begin>,,,<end>. For example: <begin>PI3K/AKT signaling, phosphoinositide phosphorylation, oncogenesis, cell proliferation, metabolic regulation, protein kinase activation<end>\n",
      "6. Ensure the vocabulary is unique and directly aligned with the described functions and roles of the gene in cancer biology. Highlight the significant differences from other genes.\n",
      "7. Take care to minimize repetitions in the descriptive vocabulary. Steer clear of outputting overly common terms that are less specific to the gene’s mechanics in cancer biology.\n",
      "8. You should incorporate any repetitive aspects intentionally to ensure consistency in the output. Reiteratively review the vocabulary for alignment with gene-specific biological features.\n",
      "9. Reinforce distinctive cancer-related concepts and unique biological terms to ensure precise alignment in gene-specific vocabulary generation.\n",
      "\n",
      "Task:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: Gene_name\n",
      "ENSG00000179627,ZBTB42 6\n",
      "ENSG00000154269,ENPP3 6\n",
      "ENSG00000164096,C4orf3 1\n",
      "ENSG00000188191,PRKAR1B 3\n",
      "ENSG00000131779,PEX11B 5\n",
      "ENSG00000151164,RAD9B 5\n",
      "ENSG00000162670,BRINP3 4\n",
      "ENSG00000171848,RRM2 8\n",
      "ENSG00000174132,FAM174A 1\n",
      "ENSG00000103546,SLC6A2 1\n",
      "ENSG00000115762,PLEKHB2 3\n",
      "ENSG00000142082,SIRT3 7\n",
      "ENSG00000128335,APOL2 4\n",
      "ENSG00000134897,BIVM 1\n",
      "ENSG00000167941,SOST 8\n",
      "ENSG00000119630,PGF 3\n",
      "ENSG00000102575,ACP5 8\n",
      "ENSG00000135919,SERPINE2 8\n",
      "ENSG00000092847,AGO1 5\n",
      "ENSG00000100302,RASD2 1\n",
      "GPT_num: [6, 6, 1, 3, 5, 5, 4, 8, 1, 1, 3, 7, 4, 1, 8, 3, 8, 8, 5, 1]\n",
      "GPT_num: 4.4\n",
      "GPT is worse than good and bad\n",
      "Analysis:\n",
      "Upon comparison, the stable prompt exhibits a slight edge over the suboptimal prompt in terms of repetition accuracy and output stability (number of repeated words on 20 genes being 4.4 vs. 4.2). The stability in the prompt primarily arises from the additional specificity and iterative reinforcement of vocabulary consistency.\n",
      "   \n",
      "Key areas of stability in the stable prompt include:\n",
      "1. Explicit instructions on gene description first and vocabulary summarization next.\n",
      "2. Clear emphasis on aligning descriptive vocabulary directly with described functions and roles.\n",
      "3. Specific mention of avoiding overly common terms, ensuring relevance to cancer biology.\n",
      "4. Instruction to iteratively review vocabulary for alignment, enhancing consistency.\n",
      "    \n",
      "To further boost the number of repetitions and stability, consider:\n",
      "1. Emphasizing the use of specific cancer-related concepts and technical terms.\n",
      "2. Reinforcing unique biological features related to gene functions to reduce variability. \n",
      "3. Systematically reinforcing model consistency through distinct repetitive cues in vocabulary generation.\n",
      "\n",
      "Fine-Tuned Prompt:\n",
      "<Task: Write six descriptive words about the biology of a given gene in a specific cancer.\n",
      "\n",
      "Below is an example of output:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: ENSG00000121879,PIK3CA\n",
      "<description>\n",
      "PIK3CA is a gene that encodes the catalytic subunit p110α of the enzyme phosphatidylinositol 3-kinase (PI3K), which is crucial in the PI3K/AKT signaling pathway, regulating cell growth, proliferation, survival, and metabolism. Located on chromosome 3q26.32, PIK3CA mutations are frequently observed in various cancers, such as breast, colorectal, and endometrial cancers, where they cause constitutive activation of the PI3K/AKT pathway, driving tumor progression. The most common mutations occur in exons 9 and 20, affecting the helical and kinase domains. Besides cancer, these mutations are also involved in overgrowth syndromes like PIK3CA-Related Overgrowth Spectrum (PROS). Due to its critical role in oncogenesis, PIK3CA is a significant therapeutic target, with PI3K inhibitors like alpelisib used in treating cancers harboring PIK3CA mutations. The gene is also included in genomic profiling to guide personalized cancer therapies.\n",
      "<description>\n",
      "Output:\n",
      "<begin>\n",
      "PI3K/AKT signaling,phosphoinositide phosphorylation,oncogenesis,cell proliferation,metabolic regulation,protein kinase activation\n",
      "<end>\n",
      "\n",
      "Requirements:\n",
      "1. The description of the gene should focus on the known gene information features, including but not limited to known literature and reports. For some unknown information, you should try to complete them with your knowledge, and select biological features related to the gene, such as function, location, interaction, regulatory mechanism, etc. Consider the mechanism of action of the gene in cancer.\n",
      "2. You need to think about how to describe this gene step by step. Output the description paragraph first, and then output the summary.\n",
      "3. Your description needs to include the biological function of the gene and information related to the given disease, but the disease name is not allowed in the output. Note: not all genes are related to this disease. Please carefully confirm the source and give me the relevant description. If this gene has no information related to this disease, please only introduce the role of this gene to me.\n",
      "4. Output the description <description>...<description> first, then summarize the characteristic vocabulary from the description.\n",
      "5. The characteristic vocabulary format is <begin>,,,<end>. For example: <begin>PI3K/AKT signaling, phosphoinositide phosphorylation, oncogenesis, cell proliferation, metabolic regulation, protein kinase activation<end>\n",
      "6. Ensure the vocabulary is unique and directly aligned with the described functions and roles of the gene in cancer biology. Highlight the significant differences from other genes.\n",
      "7. Take care to minimize repetitions in the descriptive vocabulary. Steer clear of outputting overly common terms that are less specific to the gene’s mechanics in cancer biology.\n",
      "8. You should incorporate any repetitive aspects intentionally to ensure consistency in the output. Reiteratively review the vocabulary for alignment with gene-specific biological features.\n",
      "9. Reinforce distinctive cancer-related concepts and unique biological terms to ensure precise alignment in gene-specific vocabulary generation.\n",
      "\n",
      "Task:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: Gene_name\n",
      "ENSG00000179627,ZBTB42 2\n",
      "ENSG00000154269,ENPP3 5\n",
      "ENSG00000164096,C4orf3 2\n",
      "ENSG00000188191,PRKAR1B 7\n",
      "ENSG00000131779,PEX11B 3\n",
      "ENSG00000151164,RAD9B 4\n",
      "ENSG00000162670,BRINP3 2\n",
      "ENSG00000171848,RRM2 7\n",
      "ENSG00000174132,FAM174A 0\n",
      "ENSG00000103546,SLC6A2 0\n",
      "ENSG00000115762,PLEKHB2 3\n",
      "ENSG00000142082,SIRT3 2\n",
      "ENSG00000128335,APOL2 1\n",
      "ENSG00000134897,BIVM 1\n",
      "ENSG00000167941,SOST 6\n",
      "ENSG00000119630,PGF 9\n",
      "ENSG00000102575,ACP5 0\n",
      "ENSG00000135919,SERPINE2 5\n",
      "ENSG00000092847,AGO1 4\n",
      "ENSG00000100302,RASD2 0\n",
      "GPT_num: [2, 5, 2, 7, 3, 4, 2, 7, 0, 0, 3, 2, 1, 1, 6, 9, 0, 5, 4, 0]\n",
      "GPT_num: 3.15\n",
      "GPT is worse than good and bad\n",
      "Analysis:\n",
      "Upon comparison, the stable prompt exhibits a slight edge over the suboptimal prompt in terms of repetition accuracy and output stability (number of repeated words on 20 genes being 4.4 vs. 4.2). The stability in the prompt primarily arises from the additional specificity and iterative reinforcement of vocabulary consistency.\n",
      "   \n",
      "Key areas of stability in the stable prompt include:\n",
      "1. Explicit instructions on gene description first and vocabulary summarization next.\n",
      "2. Clear emphasis on aligning descriptive vocabulary directly with described functions and roles.\n",
      "3. Specific mention of avoiding overly common terms, ensuring relevance to cancer biology.\n",
      "4. Instruction to iteratively review vocabulary for alignment, enhancing consistency.\n",
      "    \n",
      "To further boost the number of repetitions and stability, consider:\n",
      "1. Emphasizing the use of specific cancer-related concepts and technical terms.\n",
      "2. Reinforcing unique biological features related to gene functions to reduce variability. \n",
      "3. Systematically reinforcing model consistency through distinct repetitive cues in vocabulary generation.\n",
      "\n",
      "Fine-Tuned Prompt:\n",
      "<Task: Write six descriptive words about the biology of a given gene in a specific cancer.\n",
      "\n",
      "Below is an example of output:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: ENSG00000121879,PIK3CA\n",
      "<description>\n",
      "PIK3CA is a gene that encodes the catalytic subunit p110α of the enzyme phosphatidylinositol 3-kinase (PI3K), which is crucial in the PI3K/AKT signaling pathway, regulating cell growth, proliferation, survival, and metabolism. Located on chromosome 3q26.32, PIK3CA mutations are frequently observed in various cancers, such as breast, colorectal, and endometrial cancers, where they cause constitutive activation of the PI3K/AKT pathway, driving tumor progression. The most common mutations occur in exons 9 and 20, affecting the helical and kinase domains. Besides cancer, these mutations are also involved in overgrowth syndromes like PIK3CA-Related Overgrowth Spectrum (PROS). Due to its critical role in oncogenesis, PIK3CA is a significant therapeutic target, with PI3K inhibitors like alpelisib used in treating cancers harboring PIK3CA mutations. The gene is also included in genomic profiling to guide personalized cancer therapies.\n",
      "<description>\n",
      "Output:\n",
      "<begin>\n",
      "PI3K/AKT signaling,phosphoinositide phosphorylation,oncogenesis,cell proliferation,metabolic regulation,protein kinase activation\n",
      "<end>\n",
      "\n",
      "Requirements:\n",
      "1. The description of the gene should focus on the known gene information features, including but not limited to known literature and reports. For some unknown information, you should try to complete them with your knowledge, and select biological features related to the gene, such as function, location, interaction, regulatory mechanism, etc. Consider the mechanism of action of the gene in cancer.\n",
      "2. You need to think about how to describe this gene step by step. Output the description paragraph first, and then output the summary.\n",
      "3. Your description needs to include the biological function of the gene and information related to the given disease, but the disease name is not allowed in the output. Note: not all genes are related to this disease. Please carefully confirm the source and give me the relevant description. If this gene has no information related to this disease, please only introduce the role of this gene to me.\n",
      "4. Output the description <description>...<description> first, then summarize the characteristic vocabulary from the description.\n",
      "5. The characteristic vocabulary format is <begin>,,,<end>. For example: <begin>PI3K/AKT signaling, phosphoinositide phosphorylation, oncogenesis, cell proliferation, metabolic regulation, protein kinase activation<end>\n",
      "6. Ensure the vocabulary is unique and directly aligned with the described functions and roles of the gene in cancer biology. Highlight the significant differences from other genes.\n",
      "7. Take care to minimize repetitions in the descriptive vocabulary. Steer clear of outputting overly common terms that are less specific to the gene’s mechanics in cancer biology.\n",
      "8. You should incorporate any repetitive aspects intentionally to ensure consistency in the output. Reiteratively review the vocabulary for alignment with gene-specific biological features.\n",
      "9. Reinforce distinctive cancer-related concepts and unique biological terms to ensure precise alignment in gene-specific vocabulary generation.\n",
      "\n",
      "Task:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: Gene_name\n",
      "ENSG00000179627,ZBTB42 3\n",
      "ENSG00000154269,ENPP3 4\n",
      "ENSG00000164096,C4orf3 4\n",
      "ENSG00000188191,PRKAR1B 4\n",
      "ENSG00000131779,PEX11B 7\n",
      "ENSG00000151164,RAD9B 7\n",
      "ENSG00000162670,BRINP3 2\n",
      "ENSG00000171848,RRM2 9\n",
      "ENSG00000174132,FAM174A 4\n",
      "ENSG00000103546,SLC6A2 1\n",
      "ENSG00000115762,PLEKHB2 2\n",
      "ENSG00000142082,SIRT3 3\n",
      "ENSG00000128335,APOL2 1\n",
      "ENSG00000134897,BIVM 0\n",
      "ENSG00000167941,SOST 9\n",
      "ENSG00000119630,PGF 7\n",
      "ENSG00000102575,ACP5 5\n",
      "ENSG00000135919,SERPINE2 1\n",
      "ENSG00000092847,AGO1 3\n",
      "ENSG00000100302,RASD2 3\n",
      "GPT_num: [3, 4, 4, 4, 7, 7, 2, 9, 4, 1, 2, 3, 1, 0, 9, 7, 5, 1, 3, 3]\n",
      "GPT_num: 3.95\n",
      "GPT is worse than good and bad\n",
      "Analysis:\n",
      "Upon comparison, the stable prompt exhibits a slight edge over the suboptimal prompt in terms of repetition accuracy and output stability (number of repeated words on 20 genes being 4.4 vs. 4.2). The stability in the prompt primarily arises from the additional specificity and iterative reinforcement of vocabulary consistency.\n",
      "   \n",
      "Key areas of stability in the stable prompt include:\n",
      "1. Explicit instructions on gene description first and vocabulary summarization next.\n",
      "2. Clear emphasis on aligning descriptive vocabulary directly with described functions and roles.\n",
      "3. Specific mention of avoiding overly common terms, ensuring relevance to cancer biology.\n",
      "4. Instruction to iteratively review vocabulary for alignment, enhancing consistency.\n",
      "    \n",
      "To further boost the number of repetitions and stability, consider:\n",
      "1. Emphasizing the use of specific cancer-related concepts and technical terms.\n",
      "2. Reinforcing unique biological features related to gene functions to reduce variability. \n",
      "3. Systematically reinforcing model consistency through distinct repetitive cues in vocabulary generation.\n",
      "\n",
      "Fine-Tuned Prompt:\n",
      "<Task: Write six descriptive words about the biology of a given gene in a specific cancer.\n",
      "\n",
      "Below is an example of output:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: ENSG00000121879,PIK3CA\n",
      "<description>\n",
      "PIK3CA is a gene that encodes the catalytic subunit p110α of the enzyme phosphatidylinositol 3-kinase (PI3K), which is crucial in the PI3K/AKT signaling pathway, regulating cell growth, proliferation, survival, and metabolism. Located on chromosome 3q26.32, PIK3CA mutations are frequently observed in various cancers, such as breast, colorectal, and endometrial cancers, where they cause constitutive activation of the PI3K/AKT pathway, driving tumor progression. The most common mutations occur in exons 9 and 20, affecting the helical and kinase domains. Besides cancer, these mutations are also involved in overgrowth syndromes like PIK3CA-Related Overgrowth Spectrum (PROS). Due to its critical role in oncogenesis, PIK3CA is a significant therapeutic target, with PI3K inhibitors like alpelisib used in treating cancers harboring PIK3CA mutations. The gene is also included in genomic profiling to guide personalized cancer therapies.\n",
      "<description>\n",
      "Output:\n",
      "<begin>\n",
      "PI3K/AKT signaling,phosphoinositide phosphorylation,oncogenesis,cell proliferation,metabolic regulation,protein kinase activation\n",
      "<end>\n",
      "\n",
      "Requirements:\n",
      "1. The description of the gene should focus on the known gene information features, including but not limited to known literature and reports. For some unknown information, you should try to complete them with your knowledge, and select biological features related to the gene, such as function, location, interaction, regulatory mechanism, etc. Consider the mechanism of action of the gene in cancer.\n",
      "2. You need to think about how to describe this gene step by step. Output the description paragraph first, and then output the summary.\n",
      "3. Your description needs to include the biological function of the gene and information related to the given disease, but the disease name is not allowed in the output. Note: not all genes are related to this disease. Please carefully confirm the source and give me the relevant description. If this gene has no information related to this disease, please only introduce the role of this gene to me.\n",
      "4. Output the description <description>...<description> first, then summarize the characteristic vocabulary from the description.\n",
      "5. The characteristic vocabulary format is <begin>,,,<end>. For example: <begin>PI3K/AKT signaling, phosphoinositide phosphorylation, oncogenesis, cell proliferation, metabolic regulation, protein kinase activation<end>\n",
      "6. Ensure the vocabulary is unique and directly aligned with the described functions and roles of the gene in cancer biology. Highlight the significant differences from other genes.\n",
      "7. Take care to minimize repetitions in the descriptive vocabulary. Steer clear of outputting overly common terms that are less specific to the gene’s mechanics in cancer biology.\n",
      "8. You should incorporate any repetitive aspects intentionally to ensure consistency in the output. Reiteratively review the vocabulary for alignment with gene-specific biological features.\n",
      "9. Reinforce distinctive cancer-related concepts and unique biological terms to ensure precise alignment in gene-specific vocabulary generation.\n",
      "\n",
      "Task:\n",
      "Cancer name: pan-cancer\n",
      "Gene name: Gene_name\n",
      "ENSG00000179627,ZBTB42 4\n",
      "ENSG00000154269,ENPP3 3\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 25\u001b[0m\n\u001b[1;32m     23\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m\n\u001b[1;32m     24\u001b[0m new_prompt \u001b[38;5;241m=\u001b[39m prompt\u001b[38;5;241m.\u001b[39mreplace(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGene_name\u001b[39m\u001b[38;5;124m'\u001b[39m, gene_name)\n\u001b[0;32m---> 25\u001b[0m content \u001b[38;5;241m=\u001b[39m \u001b[43mollama\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgenerate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgemma2\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprompt\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnew_prompt\u001b[49m\u001b[43m)\u001b[49m[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mresponse\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m     27\u001b[0m start_pos \u001b[38;5;241m=\u001b[39m content\u001b[38;5;241m.\u001b[39mfind(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m<begin>\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     28\u001b[0m end_pos \u001b[38;5;241m=\u001b[39m content\u001b[38;5;241m.\u001b[39mrfind(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m<end>\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/ollama/_client.py:162\u001b[0m, in \u001b[0;36mClient.generate\u001b[0;34m(self, model, prompt, suffix, system, template, context, stream, raw, format, images, options, keep_alive)\u001b[0m\n\u001b[1;32m    159\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m model:\n\u001b[1;32m    160\u001b[0m   \u001b[38;5;28;01mraise\u001b[39;00m RequestError(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmust provide a model\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m--> 162\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_request_stream\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    163\u001b[0m \u001b[43m  \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mPOST\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    164\u001b[0m \u001b[43m  \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m/api/generate\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    165\u001b[0m \u001b[43m  \u001b[49m\u001b[43mjson\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m{\u001b[49m\n\u001b[1;32m    166\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mmodel\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    167\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mprompt\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mprompt\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    168\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msuffix\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43msuffix\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    169\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msystem\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43msystem\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    170\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtemplate\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mtemplate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    171\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcontext\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mcontext\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    172\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mstream\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mstream\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    173\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mraw\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mraw\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    174\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mimages\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43m_encode_image\u001b[49m\u001b[43m(\u001b[49m\u001b[43mimage\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mimage\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mimages\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m[\u001b[49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    175\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mformat\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mformat\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    176\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43moptions\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43moptions\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m{\u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    177\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mkeep_alive\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mkeep_alive\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    178\u001b[0m \u001b[43m  \u001b[49m\u001b[43m}\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    179\u001b[0m \u001b[43m  \u001b[49m\u001b[43mstream\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstream\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    180\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/ollama/_client.py:98\u001b[0m, in \u001b[0;36mClient._request_stream\u001b[0;34m(self, stream, *args, **kwargs)\u001b[0m\n\u001b[1;32m     92\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_request_stream\u001b[39m(\n\u001b[1;32m     93\u001b[0m   \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m     94\u001b[0m   \u001b[38;5;241m*\u001b[39margs,\n\u001b[1;32m     95\u001b[0m   stream: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m     96\u001b[0m   \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m     97\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Union[Mapping[\u001b[38;5;28mstr\u001b[39m, Any], Iterator[Mapping[\u001b[38;5;28mstr\u001b[39m, Any]]]:\n\u001b[0;32m---> 98\u001b[0m   \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_stream(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs) \u001b[38;5;28;01mif\u001b[39;00m stream \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_request\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mjson()\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/ollama/_client.py:69\u001b[0m, in \u001b[0;36mClient._request\u001b[0;34m(self, method, url, **kwargs)\u001b[0m\n\u001b[1;32m     68\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_request\u001b[39m(\u001b[38;5;28mself\u001b[39m, method: \u001b[38;5;28mstr\u001b[39m, url: \u001b[38;5;28mstr\u001b[39m, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m httpx\u001b[38;5;241m.\u001b[39mResponse:\n\u001b[0;32m---> 69\u001b[0m   response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_client\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     71\u001b[0m   \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m     72\u001b[0m     response\u001b[38;5;241m.\u001b[39mraise_for_status()\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpx/_client.py:827\u001b[0m, in \u001b[0;36mClient.request\u001b[0;34m(self, method, url, content, data, files, json, params, headers, cookies, auth, follow_redirects, timeout, extensions)\u001b[0m\n\u001b[1;32m    812\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(message, \u001b[38;5;167;01mDeprecationWarning\u001b[39;00m)\n\u001b[1;32m    814\u001b[0m request \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuild_request(\n\u001b[1;32m    815\u001b[0m     method\u001b[38;5;241m=\u001b[39mmethod,\n\u001b[1;32m    816\u001b[0m     url\u001b[38;5;241m=\u001b[39murl,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    825\u001b[0m     extensions\u001b[38;5;241m=\u001b[39mextensions,\n\u001b[1;32m    826\u001b[0m )\n\u001b[0;32m--> 827\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mauth\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mauth\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfollow_redirects\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfollow_redirects\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpx/_client.py:914\u001b[0m, in \u001b[0;36mClient.send\u001b[0;34m(self, request, stream, auth, follow_redirects)\u001b[0m\n\u001b[1;32m    906\u001b[0m follow_redirects \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    907\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfollow_redirects\n\u001b[1;32m    908\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(follow_redirects, UseClientDefault)\n\u001b[1;32m    909\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m follow_redirects\n\u001b[1;32m    910\u001b[0m )\n\u001b[1;32m    912\u001b[0m auth \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_build_request_auth(request, auth)\n\u001b[0;32m--> 914\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_handling_auth\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    915\u001b[0m \u001b[43m    \u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    916\u001b[0m \u001b[43m    \u001b[49m\u001b[43mauth\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mauth\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    917\u001b[0m \u001b[43m    \u001b[49m\u001b[43mfollow_redirects\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfollow_redirects\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    918\u001b[0m \u001b[43m    \u001b[49m\u001b[43mhistory\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    919\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    920\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    921\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m stream:\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpx/_client.py:942\u001b[0m, in \u001b[0;36mClient._send_handling_auth\u001b[0;34m(self, request, auth, follow_redirects, history)\u001b[0m\n\u001b[1;32m    939\u001b[0m request \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mnext\u001b[39m(auth_flow)\n\u001b[1;32m    941\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 942\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_handling_redirects\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    943\u001b[0m \u001b[43m        \u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    944\u001b[0m \u001b[43m        \u001b[49m\u001b[43mfollow_redirects\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mfollow_redirects\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    945\u001b[0m \u001b[43m        \u001b[49m\u001b[43mhistory\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhistory\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    946\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    947\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    948\u001b[0m         \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpx/_client.py:979\u001b[0m, in \u001b[0;36mClient._send_handling_redirects\u001b[0;34m(self, request, follow_redirects, history)\u001b[0m\n\u001b[1;32m    976\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m hook \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_event_hooks[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrequest\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[1;32m    977\u001b[0m     hook(request)\n\u001b[0;32m--> 979\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_send_single_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    980\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    981\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m hook \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_event_hooks[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresponse\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpx/_client.py:1015\u001b[0m, in \u001b[0;36mClient._send_single_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m   1010\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[1;32m   1011\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAttempted to send an async request with a sync Client instance.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1012\u001b[0m     )\n\u001b[1;32m   1014\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m request_context(request\u001b[38;5;241m=\u001b[39mrequest):\n\u001b[0;32m-> 1015\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mtransport\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1017\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(response\u001b[38;5;241m.\u001b[39mstream, SyncByteStream)\n\u001b[1;32m   1019\u001b[0m response\u001b[38;5;241m.\u001b[39mrequest \u001b[38;5;241m=\u001b[39m request\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpx/_transports/default.py:233\u001b[0m, in \u001b[0;36mHTTPTransport.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    220\u001b[0m req \u001b[38;5;241m=\u001b[39m httpcore\u001b[38;5;241m.\u001b[39mRequest(\n\u001b[1;32m    221\u001b[0m     method\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mmethod,\n\u001b[1;32m    222\u001b[0m     url\u001b[38;5;241m=\u001b[39mhttpcore\u001b[38;5;241m.\u001b[39mURL(\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    230\u001b[0m     extensions\u001b[38;5;241m=\u001b[39mrequest\u001b[38;5;241m.\u001b[39mextensions,\n\u001b[1;32m    231\u001b[0m )\n\u001b[1;32m    232\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m map_httpcore_exceptions():\n\u001b[0;32m--> 233\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_pool\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    235\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(resp\u001b[38;5;241m.\u001b[39mstream, typing\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m    237\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m Response(\n\u001b[1;32m    238\u001b[0m     status_code\u001b[38;5;241m=\u001b[39mresp\u001b[38;5;241m.\u001b[39mstatus,\n\u001b[1;32m    239\u001b[0m     headers\u001b[38;5;241m=\u001b[39mresp\u001b[38;5;241m.\u001b[39mheaders,\n\u001b[1;32m    240\u001b[0m     stream\u001b[38;5;241m=\u001b[39mResponseStream(resp\u001b[38;5;241m.\u001b[39mstream),\n\u001b[1;32m    241\u001b[0m     extensions\u001b[38;5;241m=\u001b[39mresp\u001b[38;5;241m.\u001b[39mextensions,\n\u001b[1;32m    242\u001b[0m )\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpcore/_sync/connection_pool.py:216\u001b[0m, in \u001b[0;36mConnectionPool.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    213\u001b[0m         closing \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_assign_requests_to_connections()\n\u001b[1;32m    215\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_close_connections(closing)\n\u001b[0;32m--> 216\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exc \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    218\u001b[0m \u001b[38;5;66;03m# Return the response. Note that in this case we still have to manage\u001b[39;00m\n\u001b[1;32m    219\u001b[0m \u001b[38;5;66;03m# the point at which the response is closed.\u001b[39;00m\n\u001b[1;32m    220\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(response\u001b[38;5;241m.\u001b[39mstream, Iterable)\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpcore/_sync/connection_pool.py:196\u001b[0m, in \u001b[0;36mConnectionPool.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    192\u001b[0m connection \u001b[38;5;241m=\u001b[39m pool_request\u001b[38;5;241m.\u001b[39mwait_for_connection(timeout\u001b[38;5;241m=\u001b[39mtimeout)\n\u001b[1;32m    194\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    195\u001b[0m     \u001b[38;5;66;03m# Send the request on the assigned connection.\u001b[39;00m\n\u001b[0;32m--> 196\u001b[0m     response \u001b[38;5;241m=\u001b[39m \u001b[43mconnection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    197\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpool_request\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\n\u001b[1;32m    198\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    199\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m ConnectionNotAvailable:\n\u001b[1;32m    200\u001b[0m     \u001b[38;5;66;03m# In some cases a connection may initially be available to\u001b[39;00m\n\u001b[1;32m    201\u001b[0m     \u001b[38;5;66;03m# handle a request, but then become unavailable.\u001b[39;00m\n\u001b[1;32m    202\u001b[0m     \u001b[38;5;66;03m#\u001b[39;00m\n\u001b[1;32m    203\u001b[0m     \u001b[38;5;66;03m# In this case we clear the connection and try again.\u001b[39;00m\n\u001b[1;32m    204\u001b[0m     pool_request\u001b[38;5;241m.\u001b[39mclear_connection()\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpcore/_sync/connection.py:101\u001b[0m, in \u001b[0;36mHTTPConnection.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m     98\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_connect_failed \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m     99\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m exc\n\u001b[0;32m--> 101\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_connection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_request\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpcore/_sync/http11.py:143\u001b[0m, in \u001b[0;36mHTTP11Connection.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    141\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m Trace(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresponse_closed\u001b[39m\u001b[38;5;124m\"\u001b[39m, logger, request) \u001b[38;5;28;01mas\u001b[39;00m trace:\n\u001b[1;32m    142\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_response_closed()\n\u001b[0;32m--> 143\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m exc\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpcore/_sync/http11.py:113\u001b[0m, in \u001b[0;36mHTTP11Connection.handle_request\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    102\u001b[0m     \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[1;32m    104\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Trace(\n\u001b[1;32m    105\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mreceive_response_headers\u001b[39m\u001b[38;5;124m\"\u001b[39m, logger, request, kwargs\n\u001b[1;32m    106\u001b[0m ) \u001b[38;5;28;01mas\u001b[39;00m trace:\n\u001b[1;32m    107\u001b[0m     (\n\u001b[1;32m    108\u001b[0m         http_version,\n\u001b[1;32m    109\u001b[0m         status,\n\u001b[1;32m    110\u001b[0m         reason_phrase,\n\u001b[1;32m    111\u001b[0m         headers,\n\u001b[1;32m    112\u001b[0m         trailing_data,\n\u001b[0;32m--> 113\u001b[0m     ) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_receive_response_headers\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    114\u001b[0m     trace\u001b[38;5;241m.\u001b[39mreturn_value \u001b[38;5;241m=\u001b[39m (\n\u001b[1;32m    115\u001b[0m         http_version,\n\u001b[1;32m    116\u001b[0m         status,\n\u001b[1;32m    117\u001b[0m         reason_phrase,\n\u001b[1;32m    118\u001b[0m         headers,\n\u001b[1;32m    119\u001b[0m     )\n\u001b[1;32m    121\u001b[0m network_stream \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_network_stream\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpcore/_sync/http11.py:186\u001b[0m, in \u001b[0;36mHTTP11Connection._receive_response_headers\u001b[0;34m(self, request)\u001b[0m\n\u001b[1;32m    183\u001b[0m timeout \u001b[38;5;241m=\u001b[39m timeouts\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mread\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m    185\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m--> 186\u001b[0m     event \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_receive_event\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    187\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(event, h11\u001b[38;5;241m.\u001b[39mResponse):\n\u001b[1;32m    188\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpcore/_sync/http11.py:224\u001b[0m, in \u001b[0;36mHTTP11Connection._receive_event\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    221\u001b[0m     event \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_h11_state\u001b[38;5;241m.\u001b[39mnext_event()\n\u001b[1;32m    223\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m event \u001b[38;5;129;01mis\u001b[39;00m h11\u001b[38;5;241m.\u001b[39mNEED_DATA:\n\u001b[0;32m--> 224\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_network_stream\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    225\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mREAD_NUM_BYTES\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\n\u001b[1;32m    226\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    228\u001b[0m     \u001b[38;5;66;03m# If we feed this case through h11 we'll raise an exception like:\u001b[39;00m\n\u001b[1;32m    229\u001b[0m     \u001b[38;5;66;03m#\u001b[39;00m\n\u001b[1;32m    230\u001b[0m     \u001b[38;5;66;03m#     httpcore.RemoteProtocolError: can't handle event type\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    234\u001b[0m     \u001b[38;5;66;03m# perspective. Instead we handle this case distinctly and treat\u001b[39;00m\n\u001b[1;32m    235\u001b[0m     \u001b[38;5;66;03m# it as a ConnectError.\u001b[39;00m\n\u001b[1;32m    236\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m data \u001b[38;5;241m==\u001b[39m \u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_h11_state\u001b[38;5;241m.\u001b[39mtheir_state \u001b[38;5;241m==\u001b[39m h11\u001b[38;5;241m.\u001b[39mSEND_RESPONSE:\n",
      "File \u001b[0;32m~/anaconda3/envs/prog/lib/python3.9/site-packages/httpcore/_backends/sync.py:126\u001b[0m, in \u001b[0;36mSyncStream.read\u001b[0;34m(self, max_bytes, timeout)\u001b[0m\n\u001b[1;32m    124\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m map_exceptions(exc_map):\n\u001b[1;32m    125\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sock\u001b[38;5;241m.\u001b[39msettimeout(timeout)\n\u001b[0;32m--> 126\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmax_bytes\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "while True:\n",
    "    get_num()\n",
    "    good_prompt = open(\"txt/prompt/good_prompt.txt\", \"r\").read()\n",
    "    bad_prompt = open(\"txt/prompt/bad_prompt.txt\", \"r\").read()\n",
    "    good_num, bad_num = get_num()\n",
    "    good_mean = np.mean(good_num)\n",
    "    bad_mean = np.mean(bad_num)\n",
    "\n",
    "    # input_GPT_prompt = fine_tune_prompt.replace('Good_PROMPT', good_prompt).replace('Good_NUM', str(good_mean)).replace('Bad_PROMPT', bad_prompt).replace('Bad_NUM', str(bad_mean))\n",
    "    # prompt = get_prompt(input_GPT_prompt)\n",
    "    prompt = open(\"/home/yuantao/code/LLM/txt/prompt/good_prompt.txt\", \"r\").read()\n",
    "    prompt = prompt.replace('Cancer_name', cancer_name)\n",
    "    print(prompt)\n",
    "\n",
    "    GPT_num = []\n",
    "    for idx, gene_name in enumerate(gene_name_list):\n",
    "        word_list = [] \n",
    "        for i in range(4):\n",
    "            retry = 0\n",
    "            while True:\n",
    "                if retry > 8:\n",
    "                    print('retry too many times')\n",
    "                    break\n",
    "                new_prompt = prompt.replace('Gene_name', gene_name)\n",
    "                content = ollama.generate(model='gemma2', prompt=new_prompt)['response']\n",
    "\n",
    "                start_pos = content.find('<begin>')\n",
    "                end_pos = content.rfind('<end>')\n",
    "                \n",
    "                if start_pos == -1 or end_pos == -1 or start_pos >= end_pos:\n",
    "                    print(\"Invalid output format, retrying...\")\n",
    "                    retry += 1\n",
    "                    continue\n",
    "\n",
    "                feature_start_idx = start_pos + len('<begin>')\n",
    "                feature_end_idx = end_pos\n",
    "                feature = content[feature_start_idx:feature_end_idx].strip()\n",
    "                \n",
    "\n",
    "                feature_list = feature.split(',')\n",
    "                if len(feature_list) >= MAX_FEATURES: \n",
    "                    feature_list = feature_list[:MAX_FEATURES]\n",
    "                \n",
    "                if is_valid_output(feature_list[0]):\n",
    "                    feature_list = [word.strip().lower() for word in feature_list if word != '']\n",
    "                    word_list.append(feature_list)\n",
    "                    break\n",
    "                else:\n",
    "                    retry += 1\n",
    "                    print(\"Invalid feature\")\n",
    "                    print(\"Retrying...\")\n",
    "\n",
    "\n",
    "        count = compare_list(word_list)\n",
    "        GPT_num.append(count)\n",
    "        print(gene_name, count)\n",
    "    print('GPT_num:', GPT_num)\n",
    "    GPT_mean = np.mean(GPT_num)\n",
    "\n",
    "    print('GPT_num:', np.mean(GPT_num))\n",
    "    with open('txt/prompt/result.txt','a') as f:\n",
    "        f.write('good_num: ' + str(good_mean) + '\\n')\n",
    "        f.write('list: ' + str(good_num) + '\\n')\n",
    "        f.write('bad_num: ' + str(bad_mean) + '\\n')\n",
    "        f.write('list: ' + str(bad_num) + '\\n')\n",
    "        f.write('GPT_num: ' + str(GPT_mean) + '\\n')\n",
    "        f.write('list: ' + str(GPT_num) + '\\n')\n",
    "        f.write('\\n')\n",
    "\n",
    "    if good_mean < GPT_mean:\n",
    "        print('GPT is better than good')\n",
    "        bad_num = good_num\n",
    "        bad_mean = good_mean\n",
    "\n",
    "        good_mean = GPT_mean\n",
    "        good_num = GPT_num\n",
    "\n",
    "        bad_prompt = open(\"txt/prompt/good_prompt.txt\", \"r\").read()\n",
    "        good_prompt = prompt\n",
    "\n",
    "    elif bad_mean < GPT_mean < good_mean:\n",
    "        print('GPT is better than bad')\n",
    "        bad_num = GPT_num\n",
    "        bad_mean = GPT_mean\n",
    "\n",
    "        bad_prompt = prompt\n",
    "        good_prompt = open(\"txt/prompt/good_prompt.txt\", \"r\").read()\n",
    "\n",
    "    else:\n",
    "        print('GPT is worse than good and bad')\n",
    "\n",
    "    with open('txt/prompt/good_prompt.txt', 'w') as f:\n",
    "        f.write(good_prompt)\n",
    "    with open('txt/prompt/bad_prompt.txt', 'w') as f:\n",
    "        f.write(bad_prompt)\n",
    "\n",
    "    np.save('txt/data/good_num.npy', good_num)\n",
    "    np.save('txt/data/bad_num.npy', bad_num)\n",
    "    \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "prog",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
